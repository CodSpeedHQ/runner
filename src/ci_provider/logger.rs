use lazy_static::lazy_static;

use crate::runner::VALGRIND_EXECUTION_TARGET;

/// This target is used exclusively to handle group events.
pub const GROUP_TARGET: &str = "codspeed::group";
pub const OPENED_GROUP_TARGET: &str = "codspeed::group::opened";

#[macro_export]
/// Start a new log group. All logs between this and the next `end_group!` will be grouped together.
///
/// # Example
///
/// ```rust
/// start_group!("My group");
/// info!("This will be grouped");
/// end_group!();
/// ```
macro_rules! start_group {
    ($name:expr) => {
        log::log!(target: $crate::ci_provider::logger::GROUP_TARGET, log::Level::Info, "{}", $name);
    };
}

#[macro_export]
/// Start a new opened log group. All logs between this and the next `end_group!` will be grouped together.
///
/// # Example
///
/// ```rust
/// start_opened_group!("My group");
/// info!("This will be grouped");
/// end_group!();
/// ```
macro_rules! start_opened_group {
    ($name:expr) => {
        log::log!(target: $crate::ci_provider::logger::OPENED_GROUP_TARGET, log::Level::Info, "{}", $name);
    };
}

#[macro_export]
/// End the current log group.
/// See [`start_group!`] for more information.
macro_rules! end_group {
    () => {
        log::log!(target: $crate::ci_provider::logger::GROUP_TARGET, log::Level::Info, "");
    };
}

pub enum GroupEvent {
    Start(String),
    StartOpened(String),
    End,
}

/// Returns the group event if the record is a group event, otherwise returns `None`.
pub(super) fn get_group_event(record: &log::Record) -> Option<GroupEvent> {
    match record.target() {
        OPENED_GROUP_TARGET => {
            let args = record.args().to_string();
            if args.is_empty() {
                None
            } else {
                Some(GroupEvent::StartOpened(args))
            }
        }
        GROUP_TARGET => {
            let args = record.args().to_string();
            if args.is_empty() {
                Some(GroupEvent::End)
            } else {
                Some(GroupEvent::Start(args))
            }
        }
        _ => None,
    }
}

lazy_static! {
    pub static ref PROVIDER_LOGGER_CONFIG: simplelog::Config = simplelog::ConfigBuilder::new()
    // We don't want to display the execution twice, so we ignore the logs from the provider target
        .add_filter_ignore_str(VALGRIND_EXECUTION_TARGET)
        .build();
}
