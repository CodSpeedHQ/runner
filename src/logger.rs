use std::{
    env,
    sync::{Arc, Mutex},
    time::Duration,
};

use console::{style, Style};
use indicatif::{ProgressBar, ProgressStyle};
use lazy_static::lazy_static;
use log::Log;
use simplelog::SharedLogger;
use std::io::Write;

pub const CODSPEED_U8_COLOR_CODE: u8 = 208; // #FF8700
const BLACK_U8_COLOR_CODE: u8 = 16; // #000

/// This target is used exclusively to handle group events.
pub const GROUP_TARGET: &str = "codspeed::group";
pub const OPENED_GROUP_TARGET: &str = "codspeed::group::opened";

#[macro_export]
/// Start a new log group. All logs between this and the next `end_group!` will be grouped together.
///
/// # Example
///
/// ```rust
/// start_group!("My group");
/// info!("This will be grouped");
/// end_group!();
/// ```
macro_rules! start_group {
    ($name:expr) => {
        log::log!(target: $crate::logger::GROUP_TARGET, log::Level::Info, "{}", $name);
    };
}

#[macro_export]
/// Start a new opened log group. All logs between this and the next `end_group!` will be grouped together.
///
/// # Example
///
/// ```rust
/// start_opened_group!("My group");
/// info!("This will be grouped");
/// end_group!();
/// ```
macro_rules! start_opened_group {
    ($name:expr) => {
        log::log!(target: $crate::logger::OPENED_GROUP_TARGET, log::Level::Info, "{}", $name);
    };
}

#[macro_export]
/// End the current log group.
/// See [`start_group!`] for more information.
macro_rules! end_group {
    () => {
        log::log!(target: $crate::logger::GROUP_TARGET, log::Level::Info, "");
    };
}

pub enum GroupEvent {
    Start(String),
    StartOpened(String),
    End,
}

/// Returns the group event if the record is a group event, otherwise returns `None`.
pub(super) fn get_group_event(record: &log::Record) -> Option<GroupEvent> {
    match record.target() {
        OPENED_GROUP_TARGET => {
            let args = record.args().to_string();
            if args.is_empty() {
                None
            } else {
                Some(GroupEvent::StartOpened(args))
            }
        }
        GROUP_TARGET => {
            let args = record.args().to_string();
            if args.is_empty() {
                Some(GroupEvent::End)
            } else {
                Some(GroupEvent::Start(args))
            }
        }
        _ => None,
    }
}

lazy_static! {
    pub static ref SPINNER: Arc<Mutex<Option<ProgressBar>>> = Arc::new(Mutex::new(None));
    pub static ref IS_TTY: bool = std::io::IsTerminal::is_terminal(&std::io::stdout());
}

/// Hide the progress bar temporarily, execute `f`, then redraw the progress bar.
///
/// If the output is not a TTY, `f` will be executed without hiding the progress bar.
pub fn suspend_progress_bar<F: FnOnce() -> R, R>(f: F) -> R {
    // If the output is a TTY, and there is a spinner, suspend it
    if *IS_TTY {
        if let Ok(mut spinner) = SPINNER.lock() {
            if let Some(spinner) = spinner.as_mut() {
                return spinner.suspend(f);
            }
        }
    }

    // Otherwise, just run the function
    f()
}

pub struct LocalLogger {
    log_level: log::LevelFilter,
}

impl LocalLogger {
    pub fn new() -> Self {
        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);

        LocalLogger { log_level }
    }
}

impl Log for LocalLogger {
    fn enabled(&self, metadata: &log::Metadata) -> bool {
        metadata.level() <= self.log_level
    }

    fn log(&self, record: &log::Record) {
        if !self.enabled(record.metadata()) {
            return;
        }

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    println!(
                        "  {}",
                        style(format!(" {} ", name.to_uppercase()))
                            .bold()
                            .color256(BLACK_U8_COLOR_CODE)
                            .on_color256(CODSPEED_U8_COLOR_CODE)
                    );
                    println!();

                    if *IS_TTY {
                        let spinner = ProgressBar::new_spinner();
                        spinner.set_style(
                            ProgressStyle::with_template(
                                format!(
                                    "  {{spinner:>.{}}} {{wide_msg:.{}.bold}}",
                                    CODSPEED_U8_COLOR_CODE, CODSPEED_U8_COLOR_CODE
                                )
                                .as_str(),
                            )
                            .unwrap(),
                        );
                        spinner.set_message(format!("{}...", name));
                        spinner.enable_steady_tick(Duration::from_millis(100));
                        SPINNER.lock().unwrap().replace(spinner);
                    } else {
                        println!("{}...", name);
                    }
                }
                GroupEvent::End => {
                    if *IS_TTY {
                        let mut spinner = SPINNER.lock().unwrap();
                        if let Some(spinner) = spinner.as_mut() {
                            spinner.finish_and_clear();
                            println!();
                        }
                    }
                    println!();
                }
            }

            return;
        }

        suspend_progress_bar(|| print_record(record));
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

/// Print a log record to the console with the appropriate style
fn print_record(record: &log::Record) {
    let error_style = Style::new().red();
    let info_style = Style::new().white();
    let warn_style = Style::new().yellow();
    let debug_style = Style::new().blue().dim();
    let trace_style = Style::new().black().dim();

    match record.level() {
        log::Level::Error => eprintln!("{}", error_style.apply_to(record.args())),
        log::Level::Warn => eprintln!("{}", warn_style.apply_to(record.args())),
        log::Level::Info => println!("{}", info_style.apply_to(record.args())),
        log::Level::Debug => println!(
            "{}",
            debug_style.apply_to(format!("[DEBUG::{}] {}", record.target(), record.args())),
        ),
        log::Level::Trace => println!(
            "{}",
            trace_style.apply_to(format!("[TRACE::{}] {}", record.target(), record.args()))
        ),
    }
}

impl SharedLogger for LocalLogger {
    fn level(&self) -> log::LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}

pub fn get_local_logger() -> Box<dyn SharedLogger> {
    Box::new(LocalLogger::new())
}

pub fn clean_logger() {
    let mut spinner = SPINNER.lock().unwrap();
    if let Some(spinner) = spinner.as_mut() {
        spinner.finish_and_clear();
    }
}
