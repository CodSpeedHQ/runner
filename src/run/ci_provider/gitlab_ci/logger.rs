use console::style;
use lazy_static::lazy_static;
use log::{Level, LevelFilter, Log};
use regex::Regex;
use simplelog::SharedLogger;
use std::{
    env,
    io::Write,
    sync::Mutex,
    time::{SystemTime, UNIX_EPOCH},
};

use crate::{
    logger::{get_group_event, GroupEvent},
    run::ci_provider::logger::should_provider_logger_handle_record,
};

lazy_static! {
    static ref GITLAB_SECTION_ID_SANITIZE_REGEX: Regex =
        Regex::new(r"[^\d\w\-_]").expect("Failed to compile GitLab SectionId regex");
}

/// Unicode Escape character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_ESC: char = '\x1B';

/// Unicode Carriage Return character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_CR: char = '\x0D';

/// Reset color mode
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
const COLOR_RESET: &str = "\x1B[0m";

/// Erase cursor till end of line
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#erase-functions
const ERASE_CURSOR: &str = "\x1B[0K";

/// A logger that prints log in the format expected by GitLab CI
///
/// See https://docs.gitlab.com/ee/ci/yaml/script.html
pub struct GitLabCILogger {
    log_level: LevelFilter,
    section_id: Mutex<Option<String>>,
}

impl GitLabCILogger {
    pub fn new() -> Self {
        // force activation of colors, because GitlabCI does not
        // respect the CLICOLORS spec.
        // https://gitlab.com/gitlab-org/gitlab/-/issues/28598
        console::set_colors_enabled(true);

        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self {
            log_level,
            section_id: Mutex::new(None),
        }
    }
}

impl Log for GitLabCILogger {
    fn enabled(&self, _metadata: &log::Metadata) -> bool {
        true
    }

    fn log(&self, record: &log::Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            let now = SystemTime::now();
            let timestamp = now.duration_since(UNIX_EPOCH).unwrap().as_secs();
            let mut section_id = self.section_id.lock().unwrap();

            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    let new_section_id = GITLAB_SECTION_ID_SANITIZE_REGEX
                        .replace_all(&name, "_")
                        .to_ascii_lowercase();

                    *section_id = Some(new_section_id.to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_start:{timestamp}:{new_section_id}{U_CR}{ERASE_CURSOR}{U_ESC}[36;1m{name}{COLOR_RESET}");
                }
                GroupEvent::End => {
                    // do not fail if there is no current section
                    let current_section_id = section_id.clone().unwrap_or("".to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_end:{timestamp}:{current_section_id}{U_CR}{ERASE_CURSOR}");

                    *section_id = None;
                }
            }
            return;
        }

        if level > self.log_level {
            return;
        }

        // set log colors. See https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
        match level {
            Level::Error => {
                println!("{}", style(message).red());
            }
            Level::Warn => {
                println!("{}", style(message).yellow());
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("{}", style(message).cyan());
            }
            Level::Trace => {
                println!("{}", style(message).magenta());
            }
        };
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GitLabCILogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
