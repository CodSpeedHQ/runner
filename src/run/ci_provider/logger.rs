use crate::run::runner::VALGRIND_EXECUTION_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the valgrind execution target
    record.target() != VALGRIND_EXECUTION_TARGET
}
