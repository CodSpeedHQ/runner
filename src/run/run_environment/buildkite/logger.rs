use crate::{
    logger::{get_group_event, get_json_event, GroupEvent},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by Buildkite
///
/// See https://buildkite.com/docs/pipelines/managing-log-output
pub struct BuildkiteLogger {
    log_level: LevelFilter,
}

impl BuildkiteLogger {
    pub fn new() -> Self {
        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self { log_level }
    }
}

impl Log for BuildkiteLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) => {
                    println!("--- {name}");
                }
                GroupEvent::StartOpened(name) => {
                    println!("+++ {name}");
                }
                GroupEvent::End => {}
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }
        // there is no support for log levels in Buildkite, so we print the level in the message
        match level {
            Level::Error => {
                println!("[ERROR] {message}");
            }
            Level::Warn => {
                println!("[WARN] {message}");
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("[DEBUG] {message}");
            }
            Level::Trace => {
                println!("[TRACE] {message}");
            }
        }
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for BuildkiteLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
