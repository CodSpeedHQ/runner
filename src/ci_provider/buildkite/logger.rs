use log::*;

use crate::ci_provider::logger::{get_group_event, GroupEvent};

/// A logger that prints logs in the format expected by Buildkite
///
/// See https://buildkite.com/docs/pipelines/managing-log-output
pub struct BuildkiteLogger;

impl Log for BuildkiteLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) => {
                    println!("--- {}", name);
                }
                GroupEvent::StartOpened(name) => {
                    println!("+++ {}", name);
                }
                GroupEvent::End => {}
            }
            return;
        }

        // there is no support for log levels in Buildkite, so we print the level in the message
        match level {
            Level::Error => {
                println!("[ERROR] {}", message);
            }
            Level::Warn => {
                println!("[WARN] {}", message);
            }
            Level::Info => {
                println!("{}", message);
            }
            Level::Debug => {
                println!("[DEBUG] {}", message);
            }
            Level::Trace => {
                println!("[TRACE] {}", message);
            }
        }
    }

    fn flush(&self) {}
}
