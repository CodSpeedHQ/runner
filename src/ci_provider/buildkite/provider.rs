use std::env;

use lazy_static::lazy_static;
use regex::Regex;

use crate::{
    ci_provider::provider::{CIProvider, CIProviderDetector},
    config::Config,
    helpers::get_env_variable,
    prelude::*,
    uploader::{RunEvent, Runner, UploadMetadata},
    VERSION,
};

use super::logger::BuildkiteLogger;

#[derive(Debug)]
pub struct BuildkiteProvider {
    pub owner: String,
    pub repository: String,
    pub ref_: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub commit_hash: String,
    pub event: RunEvent,
    pub repository_root_path: String,
}

lazy_static! {
    static ref GITHUB_URL_REGEX: Regex = Regex::new(
        r"(?x)
            (?:https://github.com/|git@github.com:)
            (?P<owner>[^/]+)/(?P<repository>[^/.]+)\.git
        "
    )
    .expect("Failed to compile GitHub URL regex");
}

pub fn get_pr_number() -> Result<Option<u64>> {
    Ok(get_env_variable("BUILDKITE_PULL_REQUEST")?.parse().ok())
}

pub fn get_run_event() -> Result<RunEvent> {
    let is_pr = get_pr_number()?.is_some();

    if is_pr {
        Ok(RunEvent::PullRequest)
    } else {
        Ok(RunEvent::Push)
    }
}

pub fn get_ref() -> Result<String> {
    let pr_number = get_pr_number()?;

    if let Some(pr_number) = pr_number {
        Ok(format!("refs/pull/{}/merge", pr_number))
    } else {
        Ok(format!(
            "refs/heads/{}",
            get_env_variable("BUILDKITE_BRANCH")?
        ))
    }
}

pub fn get_owner_and_repository() -> Result<(String, String)> {
    let repository_url = get_env_variable("BUILDKITE_REPO")?;
    let captures = GITHUB_URL_REGEX
        .captures(&repository_url)
        .context("Failed to parse the GitHub repository URL")?;

    let owner = captures
        .name("owner")
        .context("Failed to parse the GitHub repository URL")?
        .as_str();
    let repository = captures
        .name("repository")
        .context("Failed to parse the GitHub repository URL")?
        .as_str();

    Ok((owner.into(), repository.into()))
}

impl TryFrom<&Config> for BuildkiteProvider {
    type Error = Error;
    fn try_from(config: &Config) -> Result<Self> {
        if config.token.is_none() {
            bail!("Token authentication is required for Buildkite");
        }

        let is_pr = get_pr_number()?.is_some();
        let (owner, repository) = get_owner_and_repository()?;

        Ok(Self {
            owner: owner.clone(),
            repository: repository.clone(),
            ref_: get_ref()?,
            base_ref: if is_pr {
                Some(get_env_variable("BUILDKITE_PULL_REQUEST_BASE_BRANCH")?)
            } else {
                None
            },
            head_ref: if is_pr {
                Some(get_env_variable("BUILDKITE_BRANCH")?)
            } else {
                None
            },
            commit_hash: get_env_variable("BUILDKITE_COMMIT")?,
            event: get_run_event()?,
            repository_root_path: format!(
                "/buildkite/builds/{}/{}/{}/",
                get_env_variable("BUILDKITE_AGENT_NAME")?,
                get_env_variable("BUILDKITE_ORGANIZATION_SLUG")?,
                get_env_variable("BUILDKITE_PIPELINE_SLUG")?,
            ),
        })
    }
}

impl CIProviderDetector for BuildkiteProvider {
    fn detect() -> bool {
        env::var("BUILDKITE") == Ok("true".into())
    }
}

impl CIProvider for BuildkiteProvider {
    fn setup_logger(&self) -> Result<()> {
        log::set_logger(&BuildkiteLogger)?;
        Ok(())
    }

    fn get_provider_name(&self) -> &'static str {
        "Buildkite"
    }

    fn get_provider_slug(&self) -> &'static str {
        "buildkite"
    }

    fn get_upload_metadata(&self, _config: &Config, archive_hash: &str) -> Result<UploadMetadata> {
        let upload_metadata = UploadMetadata {
            base_ref: self.base_ref.clone(),
            head_ref: self.head_ref.clone(),
            commit_hash: self.commit_hash.clone(),
            event: self.event.clone(),
            owner: self.owner.clone(),
            repository: self.repository.clone(),
            ref_: self.ref_.clone(),
            repository_root_path: self.repository_root_path.clone(),

            gh_data: None,
            tokenless: false,
            // TODO: refactor in a default implementation of the trait, as it will be the same for all providers
            platform: self.get_provider_slug().into(),
            runner: Runner {
                name: "codspeed-runner".into(),
                version: VERSION.to_string(),
            },
            version: Some(1),
            profile_md5: archive_hash.to_string(),
        };

        Ok(upload_metadata)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;
    use temp_env::{with_var, with_vars};
    use url::Url;

    use super::*;

    #[test]
    fn test_detect() {
        with_var("BUILDKITE", Some("true"), || {
            assert!(BuildkiteProvider::detect());
        });
    }

    #[test]
    fn test_get_owner_and_repository() {
        with_var(
            "BUILDKITE_REPO",
            Some("https://github.com/my-org/adrien-python-test.git"),
            || {
                let (owner, repository) = get_owner_and_repository().unwrap();
                assert_eq!(owner, "my-org");
                assert_eq!(repository, "adrien-python-test");
            },
        );

        with_var(
            "BUILDKITE_REPO",
            Some("git@github.com:my-org/adrien-python-test.git"),
            || {
                let (owner, repository) = get_owner_and_repository().unwrap();
                assert_eq!(owner, "my-org");
                assert_eq!(repository, "adrien-python-test");
            },
        );
    }

    #[test]
    fn test_try_from_push_main() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("main")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("")),
                ("BUILDKITE_PULL_REQUEST", Some("")),
                (
                    "BUILDKITE_REPO",
                    Some("https://github.com/my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    command: "upload".into(),
                    skip_setup: false,
                    skip_upload: false,
                    token: Some("token".into()),
                    upload_url: Url::parse("https://example.com").unwrap(),
                    working_directory: Some(".".into()),
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();

                assert_eq!(provider.owner, "my-org");
                assert_eq!(provider.repository, "adrien-python-test");
                assert_eq!(provider.ref_, "refs/heads/main");
                assert_eq!(provider.base_ref, None);
                assert_eq!(provider.head_ref, None);
                assert_eq!(provider.commit_hash, "abc123");
                assert_eq!(provider.event, RunEvent::Push);
                assert_eq!(
                    provider.repository_root_path,
                    "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
                );
            },
        );
    }

    #[test]
    fn test_try_from_pull_request() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("feat/codspeed-runner")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("main")),
                ("BUILDKITE_PULL_REQUEST", Some("22")),
                (
                    "BUILDKITE_REPO",
                    Some("git@github.com:my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    command: "upload".into(),
                    skip_setup: false,
                    skip_upload: false,
                    token: Some("token".into()),
                    upload_url: Url::parse("https://example.com").unwrap(),
                    working_directory: Some(".".into()),
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();

                assert_eq!(provider.owner, "my-org");
                assert_eq!(provider.repository, "adrien-python-test");
                assert_eq!(provider.ref_, "refs/pull/22/merge");
                assert_eq!(provider.base_ref, Some("main".into()));
                assert_eq!(provider.head_ref, Some("feat/codspeed-runner".into()));
                assert_eq!(provider.commit_hash, "abc123");
                assert_eq!(provider.event, RunEvent::PullRequest);
                assert_eq!(
                    provider.repository_root_path,
                    "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
                );
            },
        );
    }

    #[test]
    fn test_pull_request_upload_metadata() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("feat/codspeed-runner")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("main")),
                ("BUILDKITE_PULL_REQUEST", Some("22")),
                (
                    "BUILDKITE_REPO",
                    Some("git@github.com:my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    command: "upload".into(),
                    skip_setup: false,
                    skip_upload: false,
                    token: Some("token".into()),
                    upload_url: Url::parse("https://example.com").unwrap(),
                    working_directory: Some(".".into()),
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();
                let upload_metadata = provider.get_upload_metadata(&config, "abc123").unwrap();

                assert_json_snapshot!(upload_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                })
            },
        );
    }
}
