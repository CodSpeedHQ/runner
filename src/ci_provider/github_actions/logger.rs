use crate::ci_provider::logger::{get_group_event, GroupEvent};
use log::*;
use simplelog::SharedLogger;
use std::io::Write;

/// A logger that prints logs in the format expected by GitHub Actions, with grouping support.
///
/// See https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions
pub struct GithubActionLogger;

impl Log for GithubActionLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    println!("::group::{}", name);
                }
                GroupEvent::End => {
                    println!("::endgroup::");
                }
            }
            return;
        }

        let prefix = match level {
            Level::Error => "::error::",
            Level::Warn => "::warning::",
            Level::Info => "",
            Level::Debug => "::debug::",
            Level::Trace => "::debug::[TRACE]",
        };
        let message_string = message.to_string();
        let lines = message_string.lines();
        // ensure that all the lines of the message have the prefix, otherwise GitHub Actions will not recognize the command for the whole string
        lines.for_each(|line| println!("{}{}", prefix, line));
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GithubActionLogger {
    fn level(&self) -> LevelFilter {
        // since TRACE and DEBUG use ::debug::, we always enable them and let GitHub handle the filtering
        // thanks to https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging#enabling-step-debug-logging
        LevelFilter::Trace
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
