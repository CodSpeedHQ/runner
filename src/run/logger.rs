use crate::logger::{GROUP_TARGET, OPENED_GROUP_TARGET};
use crate::prelude::*;
use crate::run::{ci_provider::CIProvider, runner::RunData};
use log::LevelFilter;
use simplelog::{CombinedLogger, WriteLogger};
use std::fs::copy;
use std::path::PathBuf;
use tempfile::NamedTempFile;

pub struct Logger {
    log_file_path: PathBuf,
}

impl Logger {
    #[allow(clippy::borrowed_box)]
    pub fn new(provider: &Box<dyn CIProvider>) -> Result<Self> {
        let provider_logger = provider.get_logger();
        let log_file = NamedTempFile::new().context("Failed to create log file")?;
        let log_file_path = log_file.path().to_path_buf();
        let file_logger_config = simplelog::ConfigBuilder::new()
            // Groups are not logged to the file
            .add_filter_ignore_str(GROUP_TARGET)
            .add_filter_ignore_str(OPENED_GROUP_TARGET)
            .build();
        let file_logger = WriteLogger::new(LevelFilter::Trace, file_logger_config, log_file);
        CombinedLogger::init(vec![provider_logger, file_logger])
            .context("Failed to init logger")?;
        Ok(Self { log_file_path })
    }

    pub fn persist_log_to_profile_folder(&self, run_data: &RunData) -> Result<()> {
        let profile_folder = run_data.profile_folder.clone();
        let dest_log_file_path = profile_folder.join("runner.log");
        debug!("Persisting log file to {}", dest_log_file_path.display());
        log::logger().flush();
        copy(&self.log_file_path, dest_log_file_path).context("Failed to copy log file")?;
        Ok(())
    }
}
