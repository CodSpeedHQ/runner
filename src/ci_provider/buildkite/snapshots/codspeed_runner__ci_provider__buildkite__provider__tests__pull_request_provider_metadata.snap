---
source: src/ci_provider/buildkite/provider.rs
expression: upload_metadata
---
{
  "ref": "refs/pull/22/merge",
  "headRef": "feat/codspeed-runner",
  "baseRef": "main",
  "owner": "my-org",
  "repository": "adrien-python-test",
  "actualCommitHash": "abc123",
  "commitHash": "abc123",
  "event": "pull_request",
  "ghData": null,
  "repositoryRootPath": "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
}
